PythonVisualizations/AdvancedSorting.py

```diff
@@ -260,7 +260,134 @@ def __quickSort(self, callEnviron, left=-1, right=-1):
+                self.set(inner, self.get(inner-h))
+                inner -= h
+                nShifts += 1
+            if inner < outer:
+                self.set(inner, temp)
+                nShifts += 1
+        h = (h - 1) // 3
+    return nShifts
+    """
+
+    def shellSort(self):
+        callEnviron = self.createCallEnvironment()
+        self.startAnimations()
+
+        # calculate h
+        h = 1
+        while h * 3 < len(self.list):
+            h = 3 * h + 1
+        hPos = (self.ARRAY_X0 + (self.CELL_WIDTH*(len(self.list)-1)) + self.CELL_WIDTH // 2, self.ARRAY_Y0-40)
+        hText = "h: {}".format(h)
+        hLabel = self.canvas.create_text(
+                *hPos, text=hText, font=self.VARIABLE_FONT,
+                fill=self.VARIABLE_COLOR)
+        callEnviron.add(hLabel)
+
+        nShifts = 0
+        while h > 0:
+            outerArrow = self.createIndex(h, name="outer", level=3)
+            callEnviron |= set(outerArrow)
+            
+            for outer in range(h, len(self.list)):
+                # move outer index
+                arrowPos = self.indexCoords(outer, level=3)
+                labelPos = (arrowPos[2], arrowPos[1])
+                self.moveItemsTo(outerArrow, (arrowPos, labelPos), sleepTime=.02)
+
+                # assign outer to temp
+                temp = self.list[outer].val
+                tempVal, label = self.assignToTemp(outer, callEnviron, varName="temp")
+                callEnviron.add(label)
+                tempAssigned = False
+
+                # create the inner index
+                inner = outer
+                innerArrow = self.createIndex(inner, name="inner", level=2)
+                callEnviron |= set(innerArrow)
+                self.wait(0.2)
+
+                # after temp is done being used, remove any number that is hidden behind another
+                toBeRemoved = []
+
+                while inner >= h and temp < self.list[inner-h].val:
+                    # get the item and value to be moved
+                    moveItems = [self.list[inner-h].display_shape]
+                    moveLoc = [self.fillCoords(self.list[inner-h].val, self.cellCoords(inner))]
+                    if self.showValues:
+                        # calculate where value should move to
+                        rectPos = self.cellCoords(inner)
+                        valPos = divide_vector(add_vector(rectPos[:2], rectPos[2:]), 2)
+                        moveItems.append(self.list[inner-h].display_val)
+                        moveLoc.append(valPos)
+                    upDelta = (0, - self.CELL_SIZE * 4 // 3)
+                    downDelta = multiply_vector(upDelta, -1)
+
+                    # item being moved should be on top
+                    for item in moveItems:
+                        self.canvas.tag_raise(item)
+                    self.moveItemsOnCurve(
+                    moveItems, moveLoc,
+                    sleepTime=0.05, startAngle=90 * 11 / (10 + abs(inner-h)))
```

Comment:
I agree that using `moveItemsOnCurve()` makes sense here for the shifting of items in the array.  When `h` is more than 1, movement on the curve makes it clearer that cells are being skipped in the assignment.  The insertionSort in SimpleSorting uses the `assignElement()` method from SortingBase for these shift assignments.   I think it would be better to give `assignElement()` an optional argument, `startAngle` that defaults to 0.   When it is 0, it uses the current `moveItemsTo()` to move them linearly, but when `startAngle` is non-zero, it would use `moveItemsOnCurve()`.   By using  `assignElement()`, we also get the copying of the canvas items and movement of the copies, which shows the state of the array cells better.  Moving this capability to SortingBase also makes it available to more array visualizations.

Reply:
Good idea! Commit 9ad42a6 addresses this